index.js

```diff
@@ -17,41 +17,56 @@ const DEFAULT_ALLOW_HEADERS = [
 
 const DEFAULT_MAX_AGE_SECONDS = 60 * 60 * 24 // 24 hours
 
-const cors = options => handler => (req, res) => {
+const checkOrigin = (origin, allowed) => {
+  if (Array.isArray(allowed)) return !!allowed.find(a => a === origin)
+  if (typeof allowed === 'function') return allowed(origin)
+  if (allowed instanceof RegExp) return allowed.test(origin)
+  return !!allowed
+}
+
+const cors = options => {
   const {
     maxAge,
-    origin,
+    origin = '*',
     allowHeaders,
     allowMethods,
   } = (options || {})
 
-  res.setHeader(
-    'Access-Control-Max-Age',
-    '' + (maxAge || DEFAULT_MAX_AGE_SECONDS)
-  )
+  return handler => (req, res) => {
 
-  res.setHeader(
-    'Access-Control-Allow-Origin',
-    (origin || '*')
-  )
+    res.setHeader(
+      'Access-Control-Max-Age',
+      '' + (maxAge || DEFAULT_MAX_AGE_SECONDS)
+    )
 
-  res.setHeader(
-    'Access-Control-Allow-Methods',
-    (allowMethods || DEFAULT_ALLOW_METHODS).join(',')
-  )
+    if (typeof origin === 'string') {
+      res.setHeader('Access-Control-Allow-Origin', origin)
+    } else if(checkOrigin(req.headers.origin, origin)) {
+      res.setHeader(
+        'Access-Control-Allow-Origin',
+        req.headers.origin
+      )
+    }
+    if (origin !== '*') res.setHeader('Vary', 'Origin')
```

Comment:
If we have a custom origin header, we'll want to let CDNs know that the Origin header needs to be preserved via this header. If we allow wildcards, this is irrelevant (and will negate some potential CDN caching).